drivers/platform/chibios/chibios_gpio.c

```diff
@@ -0,0 +1,298 @@
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#include <stdlib.h>
+#include <errno.h>
+#include "no_os_util.h"
+#include "no_os_gpio.h"
+#include "no_os_alloc.h"
+#include "chibios_gpio.h"
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Prepare the GPIO descriptor.
+ * @param desc - The GPIO descriptor.
+ * @param param - The structure that contains the GPIO parameters.
+ * @return 0 in case of success, -1 otherwise.
+ */
+static int _gpio_init(struct no_os_gpio_desc *desc,
+		      const struct no_os_gpio_init_param *param)
+{
+	struct chibios_gpio_desc *extra = desc->extra;
+	struct chibios_gpio_init_param *pextra = param->extra;
+
+	if (!param)
+		return -EINVAL;
+
+	//set GPIO pad to according mode
+	palSetPadMode(pextra->port, pextra->pad, pextra->mode);
+
+	/* copy the settings to gpio descriptor */
+	desc->port = (uint32_t) pextra->port;
+	desc->number = (uint32_t) pextra->pad;;
+	desc->pull = (((uint32_t)pextra->mode & PAL_MODE_OUTPUT_PUSHPULL)
+		      || ((uint32_t)pextra->mode & PAL_MODE_OUTPUT_OPENDRAIN)) ? NO_OS_GPIO_OUT :
+		     NO_OS_GPIO_IN;
+	extra->port = pextra->port;
+	extra->pad = pextra->pad;
+	extra->mode = pextra->mode;
+
+	return 0;
+}
+
+/**
+ * @brief Obtain the GPIO descriptor.
+ * @param desc - The GPIO descriptor.
+ * @param param - GPIO initialization parameters
+ * @return 0 in case of success, -1 otherwise.
+ */
+int chibios_gpio_get(struct no_os_gpio_desc **desc,
+		     const struct no_os_gpio_init_param *param)
+{
+	int ret;
+	struct no_os_gpio_desc *descriptor;
+	struct chibios_gpio_desc *extra;
+
+	if (!desc || !param)
+		return -EINVAL;
+
+	descriptor = (struct no_os_gpio_desc *)no_os_malloc(sizeof(
+				struct no_os_gpio_desc));
+	if (!descriptor) {
+		ret = -ENOMEM;
+		goto error_1;
+	}
+	extra = (struct chibios_gpio_desc*)no_os_malloc(sizeof(struct
+			chibios_gpio_desc));
+	if (!extra) {
+		ret = -ENOMEM;
+		goto error_2;
+	}
+
+	descriptor->extra = extra;
+	ret = _gpio_init(descriptor, param);
+	if (ret < 0)
```

Comment:
nit: similar to the approach on linux, we prefer checking `if (ret)` if we don't have any positive values that have a special meaning.